search-app/src/cmr/search/api/routes.clj

```diff
@@ -231,14 +238,44 @@
         results (query-svc/find-concepts-by-parameters ctx concept-type search-params)]
     (search-response ctx results)))
 
+(defn- find-granule-parent-collections
+  "Invokes query service to find collections based on data found in a granule search,
+   then executes a search for collections with the found concept-ids. Supports CMR Harvesting."
+  [ctx path-w-extension params headers body]
+  (let [concept-type (concept-type-path-w-extension->concept-type path-w-extension)
+        ctx (assoc ctx :query-string body)
+        params (process-params params path-w-extension headers mt/xml)
+        collections-with-new-granules (query-svc/new-granules->collection-ids ctx params)
+        search-params (-> params
+                          (assoc :echo-collection-id collections-with-new-granules)
+                          (dissoc :has-granules-created-at)
+                          lp/process-legacy-psa)]
+    (find-concepts-by-parameters ctx path-w-extension search-params headers body)))
+
+(defn- multi-part-query?
+  "Return true if parameters match any from the list of `multi-part-query-params`
+   defined above. Supports CMR Harvesting."
+  [search-parameters]
+  (some true? (map #(contains? search-parameters %) multi-part-query-params)))
+
 (defn- find-concepts
-  "Invokes query service to find results and returns the response"
+  "Invokes query service to find results and returns the response.
+
+  This function supports several approaches/cases/whatevs for obtaining concept data:
+  * By JSON query
+  * By parameter string and URL parameters
+  * Collections from Granules - due to the fact that ES doesn't suport joins in the way
+    that we need, we have to make two queires here to support CMR Harvesting. This can
+    later be generalized easily, should the need arise."
   [ctx path-w-extension params headers body]
   (let [content-type-header (get headers (string/lower-case common-routes/CONTENT_TYPE_HEADER))]
     (cond
       (= mt/json content-type-header)
       (find-concepts-by-json-query ctx path-w-extension params headers body)
 
+      (multi-part-query? params)
```

Comment:
I appreciate you wanting to leave room to generalize this later, but right now I find it a bit confusing that you have a predicate checking if it's a multipart query jumping straight to `find-granule-parent-collections`. It would make more sense to me if you stopped trying to generalize it and just called the predicate `granule-parent-collection-query?` or if you adde a level of indirection here with an intermediate function like `find-concepts-with-multi-query` that conditionally calls the right function. For now the condition would always call `find-granule-parent-collections`.